### PR TITLE
Add logging infra to PineconeClient and Index.

### LINKED ISSUE #118: Consider integrating `ILogger` into `PineconeClient` and `ITransport<T>` implementations
For upcoming (and not only) SK integration, it might be useful to allow for flowing down `ILogger` instance and either using it internally or further passing it to `HttpClient`/`GrpcChannel` where appropriate.

### PR DESCRIPTION
**Pull request by maumar**
The idea is, each operation has a pair of log entries - start is logged at Trace level (and should be ignored in most cases), and then we either log success at Debug level, or error at Error level. Only logging basic information like operation type and the index name - we want to avoid printing sensitive data in the logs for security reasons.

Also wired up logger(factory) to GrpcChannel and HttpClient for more detailed logging.

Fixes #118

### CODE REVIEW

File: src/PineconeClient.cs [hunk 1]
```diff
@@ -28,12 +34,17 @@ public PineconeClient(string apiKey)
...
+    public PineconeClient(string apiKey, Uri baseUrl, ILoggerFactory? loggerFactory = null)
     {
         Guard.IsNotNullOrWhiteSpace(apiKey);
         Guard.IsNotNull(baseUrl);
 
-        Http = new() { BaseAddress = baseUrl };
+        LoggerFactory = loggerFactory;
+        Logger = loggerFactory?.CreateLogger<PineconeClient>() ?? (ILogger)NullLogger.Instance;
+
+        var httpClientLogger = loggerFactory?.CreateLogger<HttpClient>() ?? (ILogger)NullLogger.Instance;
```
Comment (maumar) [inline src/PineconeClient.cs:46]: not sure if this is the right way to do this

**Comment by neon-sunset** [comment]
It looks like I'm down with COVID or similar and will get to this in a couple days. Thanks!

**Comment by maumar** [comment]
@neon-sunset get well soon! I'm planning to take a look at rom<float> improvement as well and can do that in parallel to this work, so no rush.

**Comment by neon-sunset** [comment]
@maumar I have investigated the issue with credentials leakage as discussed and could not reproduce it with gRPC stack - it is generally well-behaved if, as you noted, chatty, and does not output sensitive details about payload or metadata.

Regarding explicit logging within client and transports/index - I'm slightly on the fence about it. When looking at it without context the logging from purely gRPC and HttpClient themselves seems insufficient. At the same time, I have quite a bit of experience doing incident response and L3 support where these extra "Operation Started"+"Operation Completed/Failed" pairs have only ever added to visual noise when looking through logs and increased log storage bill - most of the time, you only ever care about innermost failure message that contains exception details, and we're bubbling up exceptions to the caller therefore this will end up being eventually observed and logged by users who care about this.

For now, I decided to try out a sample application where only "interesting" parts are logged, which is only one thing for now - parallel batch operations.

While stress-testing these I discovered that Pinecone has quite stringent payload size limits, so I want to flesh this feature out a bit more before doing a release - there's a 1k IDs limit on Delete request, so it must have parallel batching too.

Regarding official/Fern's Pinecone SDK at https://github.com/pinecone-io/pinecone-dotnet-client/https://www.nuget.org/packages/Pinecone.Client - it was indeed released. However, for now it has a few issues which I believe make it worth continuing to develop Pinecone.NET:

- blocking operations where users do not expect them: https://github.com/pinecone-io/pinecone-dotnet-client/blob/main/src/Pinecone/PineconeClient.cs#L28
- array allocations (up to 12KiB for 3072 dimensions) https://github.com/pinecone-io/pinecone-dotnet-client/blob/main/src/Pinecone/Types/Vector.cs#L67 avoiding which requires more opinionated if riskier approach
- only gRPC transport is available - being able to only use REST is still relevant to environments which restrict HTTP protocols in use, but it is also relevant to applications that want to minimize binary size because using REST relies on the same code referenced code by `PineconeClient`'s base operations, reducing size impact by 2-2.5MiB. I have also discovered that gRPC has lower throughput than REST from my location to PoP in US when doing parallel fetch/upserts
- despite using gRPC, Pinecone.Client is not AOT-compatible

It is, however, more customizable w.r.t. timeout and retries configuration but this can be addressed if current defaults pose issues to end users.

**Comment by neon-sunset** [comment]
I think it can be merged in the current form, I have also added a redact options to logging message handler so that the API key is now filtered regardless of the logging level. In any case, thank you for pointing this out previously and for contributing this PR.

**Comment by neon-sunset** [comment]
@maumar The version 3.0.0 which includes this change has been released: https://github.com/neon-sunset/Pinecone.NET/releases
This release also switches to `ReadOnlyMemory<float>` to stop forcing Semantic Kernel to allocate large arrays whenever constructing vectors for Pinecone. The mapping in SK needs to be updated to account for this, however.

Thank you!

**Pull request merged**